src/store/eventing/middleware.ts

```diff
@@ -0,0 +1,30 @@
+import { Action, Dispatch, Middleware, MiddlewareAPI } from '@reduxjs/toolkit';
+import { toggleAnnotationModeAction } from '../common/actions';
+import handleModeEvents from './modes';
+import handleVisibilityEvents from './visibility';
+import { setVisibilityAction } from '../common';
+import { EventHandlerEntry } from './types';
+
+// Array of event handlers based on redux action. To add handling for new events add an entry keyed by action
+const eventHandlers: EventHandlerEntry[] = [
+    { [toggleAnnotationModeAction.toString()]: handleModeEvents },
```

Comment:
I'm wondering if we can compare the objects directly if we use a WeakMap rather than an array of objects? It doesn't seem likely that we'll need multiple handlers for the same action.

```
const eventHandlers = new WeakMap();

eventHandlers.set(setVisibilityAction, handleVisibilityEvents);
eventHandlers.set(toggleAnnotationModeAction, handleModeEvents);

//...later
eventHandlers.get(action)(prevState, nextState)

Reply:
I tried this out, but I'm not sure that it works because the action that is the key into the WeakMap isn't the same as the action received by the middleware